actors/builtin/miner/miner_actor.go

```diff
@@ -1653,141 +1653,59 @@ func handleProvingDeadline(rt Runtime) {
-			}
-
-			depositToBurn, err := st.checkPrecommitExpiry(store, bf)
+			depositToBurn, err := st.ExpirePreCommits(store, currEpoch)
 			builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to expire pre-committed sectors")
-			penaltyTotal = big.Add(penaltyTotal, depositToBurn)
+			penaltyTarget = big.Add(penaltyTarget, depositToBurn)
 		}
 
 		// Record whether or not we _had_ early terminations in the queue before this method.
 		// That way, don't re-schedule a cron callback if one is already scheduled.
 		hadEarlyTerminations = havePendingEarlyTerminations(rt, &st)
 
-		// Note: because the cron actor is not invoked on epochs with empty tipsets, the current epoch is not necessarily
-		// exactly the final epoch of the deadline; it may be slightly later (i.e. in the subsequent deadline/period).
-		// Further, this method is invoked once *before* the first proving period starts, after the actor is first
-		// constructed; this is detected by !dlInfo.PeriodStarted().
-		// Use dlInfo.PeriodEnd() rather than rt.CurrEpoch unless certain of the desired semantics.
-		dlInfo := st.DeadlineInfo(currEpoch)
-		if !dlInfo.PeriodStarted() {
-			return // Skip checking faults on the first, incomplete period.
-		}
-		deadlines, err := st.LoadDeadlines(store)
-		builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to load deadlines")
-		deadline, err := deadlines.LoadDeadline(store, dlInfo.Index)
-		builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to load deadline %d", dlInfo.Index)
-		quant := dlInfo.QuantSpec()
-		unlockedBalance := st.GetUnlockedBalance(rt.CurrentBalance())
-
 		{
-			// Detect and penalize missing proofs.
-			faultExpiration := dlInfo.Last() + FaultMaxAge
-
-			partPowerDelta, penalizedPower, err := deadline.ProcessDeadlineEnd(store, quant, faultExpiration)
-			builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to process end of deadline %d", dlInfo.Index)
-
-			powerDelta = powerDelta.Add(partPowerDelta)
-
-			// Unlock sector penalty for all undeclared faults.
-			penaltyTarget := PledgePenaltyForUndeclaredFault(epochReward.ThisEpochRewardSmoothed, pwrTotal.QualityAdjPowerSmoothed, penalizedPower.QA)
-			// Subtract the "ongoing" fault fee from the amount charged now, since it will be added on just below.
-			penaltyTarget = big.Sub(penaltyTarget, PledgePenaltyForDeclaredFault(epochReward.ThisEpochRewardSmoothed, pwrTotal.QualityAdjPowerSmoothed, penalizedPower.QA))
-			penaltyFromVesting, penaltyFromBalance, err := st.PenalizeFundsInPriorityOrder(store, currEpoch, penaltyTarget, unlockedBalance)
-			builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to unlock penalty")
-			unlockedBalance = big.Sub(unlockedBalance, penaltyFromBalance)
-			penaltyTotal = big.Sum(penaltyTotal, penaltyFromVesting, penaltyFromBalance)
-			pledgeDelta = big.Sub(pledgeDelta, penaltyFromVesting)
+			penalty, pledgeDelta, powerDelta, err := st.AdvanceDeadline(
```

Comment:
This will short-circuit if:

* The period hasn't started.
* There is no "initial pledge" (no sectors). Can we assume that every sector will have a non-zero initial pledge?
* We load the deadline and find it has no sectors.

Reply:
As for PCD, I am not comfortable coupling a test for the existence of any sectors to the presence of pledge.

Reply:
IMO this has gone too far in pushing things down into the state. I do like the manipulations being pushed there, but do not think that "policy" logic like how much penalty to pay should be there.

Reply:
SGTM.

Reply:
Yeah, you're right. I've moved fee calculation back up.